### PR TITLE
Ees 4683 2

### PR DESCRIPTION
**Pull request by Dancemammal**
This pull request covers the git comments from the old pull request, it has been changed to remove some of the blockers to deploying on to the DfE subscription.
Unused resources have been removed (blobstore)
Some resources have been merged together to save complexity
Deployment flags have been added to cover some resources deployment
Parameters have been provided for Dev, Test, Pre-prod and Production
Resource naming now follows DfE standards more closely

### CODE REVIEW

**Review by ntsim — CHANGES_REQUESTED** [review]

File: infrastructure/templates/public-api/main.bicep [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+@description('Specifies the container port.')
+param targetPort int = 80
+
+@description('Select if you want to use a public dummy image to start the container app.')
+param useDummyImage bool
+
+
+//ServiceBus Queue Params -------------------------------------------------------------------
+@description('Name of the Service Bus namespace')
+param namespaceName string = 'processor'
```
Comment (ntsim) [inline infrastructure/templates/public-api/main.bicep:115]: Would suggest being a bit more explicit and naming this `publicapi-processor`

File: infrastructure/templates/public-api/main.bicep [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+@description('Name of the Queue')
+param queueName string = 'Processorqueue'
```
Comment (ntsim) [inline infrastructure/templates/public-api/main.bicep:118]: This doesn't follow the normal naming conventions. Would suggest `publicapi-processor`

File: infrastructure/templates/public-api/main.bicep [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+//ETL Function Paramenters ------------------------------------------------------------------
+@description('Specifies the name of the function.')
+param functionAppName string = 'processor'
```
Comment (ntsim) [inline infrastructure/templates/public-api/main.bicep:122]: Similar to other comments, would go with `publicapi-processor`

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 4]
```diff
@@ -0,0 +1,199 @@
+# Infrastructure 
+# Check and build infrastructure
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:2]: Removed

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  isMain: $[eq(variables['Build.SourceBranch'], 'refs/heads/main')]
+  vmImageName: 'ubuntu-latest'
+  resourceGroupName: 'dfe-development'
+  location: 'uksouth'
+  azureSubscription: 'AzureConnection'
+  templateFile: '/public-api/main.bicep'
+  devtemplateParams: '/public-api/parameters/main-dev.bicepparam'
+  testtemplateParams: '/public-api/parameters/main-test.bicepparam'
+  preprodtemplateParams: '/public-api/parameters/main-preprod.bicepparam'
+  prodtemplateParams: '/public-api/parameters/main-prod.bicepparam'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:28]: Updated

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+pool:
+  vmImage: $(vmImageName)
+
+stages:
+- stage: Build
+  jobs:
+  - job: Check_Infrastructure
+    steps:
+    ####### Check Infrastructure ########
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:38]: Removed

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    - checkout: self
+    - task: CmdLine@2
+      displayName: 'List source files'
+      inputs:
+        script: |
+          echo "Structure of work folder of this pipeline:"
+          tree $(Build.SourcesDirectory)
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:45]: Removed

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        azureSubscription: $(azureSubscription) 
+        scriptType: 'bash' # You could also use Batch or PowerShell if desired
+        scriptLocation: 'inlineScript'
+        inlineScript: 'az bicep install'
+
+    - task: AzureCLI@2
+      displayName: 'Validate bicep scripts' 
+      inputs:
+        azureSubscription: $(azureSubscription) 
+        scriptType: 'pscore'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:59]: Changed to Bash

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    - infrastructure/public-api/*
+pr: none
+
+name: Deploy Public API Bicep files
+
+variables:
+  isDev: $[eq(variables['Build.SourceBranch'], 'refs/heads/dev')]
+  isMain: $[eq(variables['Build.SourceBranch'], 'refs/heads/main')]
+  vmImageName: 'ubuntu-latest'
+  resourceGroupName: 'dfe-development'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:21]: Updated to be a parameter

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+  location: 'uksouth'
+  azureSubscription: 'AzureConnection'
+  templateFile: '/public-api/main.bicep'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:24]: Moved file to public-API

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+        scriptLocation: 'inlineScript'
+        inlineScript: 'az bicep install'
+
+    - task: AzureCLI@2
+      displayName: 'Validate bicep scripts' 
+      inputs:
+        azureSubscription: $(azureSubscription) 
+        scriptType: 'pscore'
+        scriptLocation: 'inlineScript'
+        inlineScript: 'az deployment group validate --resource-group $(resourceGroupName) --template-file $(Build.SourcesDirectory)$(templateFile) --parameters $(Build.SourcesDirectory)$(devtemplateParams)'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:61]: updated

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+name: Deploy Public API Bicep files
+
+variables:
+  isDev: $[eq(variables['Build.SourceBranch'], 'refs/heads/dev')]
+  isMain: $[eq(variables['Build.SourceBranch'], 'refs/heads/main')]
+  vmImageName: 'ubuntu-latest'
+  resourceGroupName: 'dfe-development'
+  location: 'uksouth'
+  azureSubscription: 'AzureConnection'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:23]: updated to parameter

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+  preprodtemplateParams: '/public-api/parameters/main-preprod.bicepparam'
+  prodtemplateParams: '/public-api/parameters/main-prod.bicepparam'
+
+pool:
+  vmImage: $(vmImageName)
+
+stages:
+- stage: Build
+  jobs:
+  - job: Check_Infrastructure
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:36]: changed

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+      inputs:
+        azureSubscription: $(azureSubscription) 
+        scriptType: 'pscore'  
+        scriptLocation: 'inlineScript'  
+        inlineScript: 'az bicep build --file $(Build.SourcesDirectory)$(templateFile)'  
+
+    - task: PublishBuildArtifacts@1
+      displayName: 'Publish json artifact in pipeline'
+      inputs:
+        PathtoPublish: '$(Build.SourcesDirectory)'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:74]: Changed

File: infrastructure/templates/public-api/components/acrSeeder.bicep [hunk 15]
```diff
@@ -0,0 +1,51 @@
+@description('Specifies the Resource Prefix')
+param resourcePrefix string
+
+@description('Specifies the location for all resources.')
+param location string
+
+@minLength(5)
+@maxLength(50)
+@description('Name of the azure container registry (must be globally unique)')
+param containerRegistryName string = 'eesapiacr'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/acrSeeder.bicep:10]: Removed default

File: infrastructure/templates/public-api/components/containerApp.bicep [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+  '0.75'
+  '1'
+  '1.25'
+  '1.5'
+  '1.75'
+  '2'
+])
+param cpuCore string = '0.5'
+
+@description('Amount of memory (in gibibytes, GiB) allocated to the container up to 4GiB. Can be with a maximum of two decimals. Ratio with CPU cores must be equal to 2.')
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/containerApp.bicep:44]: Not a typo, text is from Microsoft: https://learn.microsoft.com/en-us/azure/container-apps/containers

File: infrastructure/templates/public-api/components/containerApp.bicep [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+@allowed([
+  '0.5'
+  '1'
+  '1.5'
+  '2'
+  '3'
+  '3.5'
+  '4'
+])
+param memorySize string = '1'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/containerApp.bicep:54]: Changed environment to handle D4 Workload Profile

File: infrastructure/templates/public-api/parameters/main-dev.bicepparam [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+param domain = 'publicapi'
+param subscription = 's101d01'
+param environment = 'eesapi'
+param environmentName = 'Development'
+
+//Networking Params --------------------------------------------------------------------------
+param deploySubnets = true
+
+//PostgreSQL Database Params -------------------------------------------------------------------
+param postgreSQLserverName = 'metadata'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/parameters/main-dev.bicepparam:13]: updated and moved to main

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+        scriptLocation: 'inlineScript'
+        inlineScript: 'az deployment group validate --resource-group $(resourceGroupName) --template-file $(Build.SourcesDirectory)$(templateFile) --parameters $(Build.SourcesDirectory)$(devtemplateParams)'
+        
+    - task: AzureCLI@2  
+      displayName: 'Build bicep artifact for testing' 
+      inputs:
+        azureSubscription: $(azureSubscription) 
+        scriptType: 'pscore'  
+        scriptLocation: 'inlineScript'  
+        inlineScript: 'az bicep build --file $(Build.SourcesDirectory)$(templateFile)'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:69]: changed

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+        azureSubscription: $(azureSubscription) 
+        scriptType: 'pscore'  
+        scriptLocation: 'inlineScript'  
+        inlineScript: 'az bicep build --file $(Build.SourcesDirectory)$(templateFile)'  
+
+    - task: PublishBuildArtifacts@1
+      displayName: 'Publish json artifact in pipeline'
+      inputs:
+        PathtoPublish: '$(Build.SourcesDirectory)'  
+        ArtifactName: 'finishedTemplate'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:75]: Section removed

File: infrastructure/templates/api-infrastructure-pipeline.yml [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+pool:
+  vmImage: $(vmImageName)
+
+stages:
+- stage: Build
+  jobs:
+  - job: Check_Infrastructure
+    steps:
+    ####### Check Infrastructure ########
+    - checkout: self
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/api-infrastructure-pipeline.yml:39]: Best practice, and will benefit others reading the script

File: infrastructure/templates/public-api/components/acrSeeder.bicep [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+@description('Specifies the base docker container image to deploy.')
+param containerSeedImage string = 'mcr.microsoft.com/azuredocs/aci-helloworld'
+
+//Variables 
+var containerImportName = '${resourcePrefix}ImportContainerImage'
+var userIdentityName = '${resourcePrefix}-id-seeder'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/acrSeeder.bicep:17]: remove seeder

File: infrastructure/templates/public-api/application/etlFunctionApp.bicep [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+
+@description('Specifies the location for all resources.')
+param location string
+
+//Specific parameters for the resources
+@description('Function App : Runtime Language')
+param functionAppRuntime string = 'dotnet'
+
+@description('Specifies the name of the function.')
+param functionAppName string = 'processor'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/application/etlFunctionApp.bicep:12]: updated

File: infrastructure/templates/public-api/components/acrSeeder.bicep [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+//Resources 
+
+//Managed Identity
+resource managedIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' = {
+  name: userIdentityName
+  location: location
+}
+
+@description('This allows the managed identity of the seeder to access the registry, note scope is applied to the wider ResourceGroup not the ACR')
+resource managedIdentityRBAC 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/acrSeeder.bicep:29]: We can create the managed identity with the correct name, we need DfE to add the correct RBAC role and that will work

[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/acrSeeder.bicep:29]: ACR Seeder removed

File: infrastructure/templates/public-api/components/appInsights.bicep [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+@description('Application Insights name')
+param appInsightsName string
+
+// Variables and created data
+var kind = 'web'
+
+
+//Resources
+resource applicationInsights 'Microsoft.Insights/components@2020-02-02' = {
+  name: appInsightsName
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/appInsights.bicep:15]: Updated

File: infrastructure/templates/public-api/parameters/main-dev.bicepparam [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+param dbAdminName = 'PostgreSQLAdmin'
+param dbAdminPassword = 'postgreSQLAdminPassword'
+param dbSkuName = 'Standard_B1ms'
+param storageSizeGB = 32
+param autoGrowStatus = 'Disabled'
+
+//Container Registry -------------------------------------------------------------------
+param containerRegistryName = 'eesapiacr'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/parameters/main-dev.bicepparam:21]: changed

File: infrastructure/templates/public-api/main.bicep (hunk 3 above)
Comment (Dancemammal) [reply infrastructure/templates/public-api/main.bicep:122]: done

File: infrastructure/templates/public-api/components/functionApp.bicep [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+  properties: {
+    AzureWebJobsStorage: storageAccountConnectionString
+    WEBSITE_CONTENTAZUREFILECONNECTIONSTRING: storageAccountConnectionString
+    WEBSITE_CONTENTSHARE: toLower(functionAppName)
+    FUNCTIONS_EXTENSION_VERSION: '~4'
+    APPINSIGHTS_INSTRUMENTATIONKEY: applicationInsightsModule.outputs.applicationInsightsKey
+    FUNCTIONS_WORKER_RUNTIME: functionAppRuntime
+    WEBSITE_RUN_FROM_PACKAGE: '1'
+    DatabaseConnectionString: databaseConnectionStringKeyVaultRef
+    ServiceBusConnectionString: serviceBusConnectionString
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/functionApp.bicep:109]: done

File: infrastructure/templates/public-api/components/keyVaultAccessPolicy.bicep [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+@description('Specifies the tenant ID of the resource that should have access to the Key Vault')
+param tenantId string
+
+
+// Variables and created data
+
+
+//Resources
+resource keyVaultAccessPolicy 'Microsoft.KeyVault/vaults/accessPolicies@2022-07-01' = {
+  name: '${keyVaultName}/add'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/keyVaultAccessPolicy.bicep:17]: MS states this is the convention to add a new policy

File: infrastructure/templates/public-api/main.bicep [hunk 29]
```diff
@@ -0,0 +1,301 @@
...
+param containerRegistryName string
+
+@description('Deploy the Container Registry if you are not using an existing registry')
+param deployRegistry bool
+
+@description('Container App : Specifies the container image to seed to the ACR.')
+param containerSeedImage string
+
+@description('Select if you want to seed the ACR with a base image.')
+param seedRegistry bool
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/main.bicep:89]: removed

File: infrastructure/templates/public-api/main.bicep [hunk 30]
```diff
@@ -0,0 +1,301 @@
...
+    resourcePrefix: resourcePrefix
+    location: location
+    containerRegistryName: containerRegistryModule.outputs.containerRegistryName
+    containerSeedImage: containerSeedImage
+  }
+  dependsOn: [
+    containerRegistryModule
+    keyVaultModule
+  ]
+}
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/main.bicep:236]: removed

File: infrastructure/templates/public-api/parameters/main-dev.bicepparam [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+param environment = 'eesapi'
+param environmentName = 'Development'
+
+//Networking Params --------------------------------------------------------------------------
+param deploySubnets = true
+
+//PostgreSQL Database Params -------------------------------------------------------------------
+param postgreSQLserverName = 'metadata'
+param dbAdminName = 'PostgreSQLAdmin'
+param dbAdminPassword = 'postgreSQLAdminPassword'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/parameters/main-dev.bicepparam:15]: now a parameter

File: infrastructure/templates/public-api/components/storageaccount.bicep [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+        action: 'Allow'
+      }]
+    }
+  }
+  tags: tagValues
+}
+
+//Outputs
+output storageAccountName string = storageAccount.name
+output storageAccountConnectionString string = 'DefaultEndpointsProtocol=https;AccountName=${storageAccount.name};EndpointSuffix=${endpointSuffix};AccountKey=${key}'
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/storageaccount.bicep:67]: complete

File: infrastructure/templates/public-api/components/postgresqlDatabase.bicep [hunk 33]
```diff
@@ -0,0 +1,114 @@
...
+
+
+//Outputs
+@description('The fully qualified Azure resource ID of the Database Server.')
+output databaseRef string = resourceId('Microsoft.DBforPostgreSQL/flexibleServers', databaseName)
+
+@description('Connection String Secrets.')
+output connectionStringSecretName string = connectionStringSecretName
+output connectionStringSecretUri string = storeADOConnectionStringToKeyVault.outputs.keyVaultSecretUri
+output dbConnectionString string = connectionString
```
[the remark this answers is not in the document]
Comment (Dancemammal) [reply infrastructure/templates/public-api/components/postgresqlDatabase.bicep:114]: complete

**Pull request closed without merging**